### PR TITLE
Add dependencies route

### LINKED ISSUE #97: Gem dependencies API
Let's investigate what is `/api/v1/dependencies` API and create an SDK method to cover this API request.

### PR DESCRIPTION
**Pull request by milkysoft**
1. Implement /dependencies route
2. Implement /versions route
3. Implement /gems route
4. Make `bundle install` command to succeed

close #97

### CODE REVIEW

**Comment by g4s8** [comment]
@milkysoft also, please pay attention to PR style: https://github.com/artipie/artipie/blob/master/CONTRIBUTING.md

File: src/main/java/com/artipie/gem/Gem.java [hunk 1]
```diff
@@ -89,6 +96,71 @@ public Gem(final Storage storage) {
         this.shared = new SharedRuntime();
     }
 
+    /**
+     * Get info Ruby gem.
```
Comment (g4s8) [inline src/main/java/com/artipie/gem/Gem.java:100]: Why do we need to load Ruby gem into memory?

File: src/main/java/com/artipie/gem/Gem.java [hunk 2]
```diff
@@ -89,6 +96,71 @@ public Gem(final Storage storage) {
...
+     *
+     * @param filename Ruby file to return
+     * @return Completable action
+     */
+    public CompletionStage<byte[]> getRubyFile(final Key filename) {
```
Comment (g4s8) [inline src/main/java/com/artipie/gem/Gem.java:105]: Key is not a file

File: src/main/java/com/artipie/gem/Gem.java [hunk 3]
```diff
@@ -89,6 +96,71 @@ public Gem(final Storage storage) {
...
+        )
+        .thenApply(
+            tmpdir -> {
+                final byte[] filecontent;
+                try {
+                    final CompletionStage<Key> stage = this.getGemFile(
+                        filename, true, new Key.From("thor-1.0.1.gemspec.rz")
+                    );
+                    final CompletableFuture<Key> feature = stage.toCompletableFuture();
+                    final Key thekey = feature.get(10, TimeUnit.MILLISECONDS);
```
Comment (g4s8) [inline src/main/java/com/artipie/gem/Gem.java:123]: Please don't use blocking calls, vertx server may stop the request processor on such behavior

File: src/main/java/com/artipie/gem/Gem.java [hunk 4]
```diff
@@ -89,6 +96,71 @@ public Gem(final Storage storage) {
...
+                    final Path path = Paths.get(tmpdir.toString(), thekey.string());
+                    final File file = new File(path.toString());
+                    filecontent = Files.readAllBytes(file.toPath());
```
Comment (g4s8) [inline src/main/java/com/artipie/gem/Gem.java:126]: Why do you convert path to file and then back to path? You can just use `Files.readAllBytes(path)`

File: src/main/java/com/artipie/gem/Gem.java [hunk 5]
```diff
@@ -89,6 +96,71 @@ public Gem(final Storage storage) {
...
+                } catch (final IOException | TimeoutException | InterruptedException
```
Comment (g4s8) [inline src/main/java/com/artipie/gem/Gem.java:127]: `InterruptedException` should not be handled in a such way. Please read the documentation about correct `InterruptedException` handling

File: src/main/java/com/artipie/gem/Gem.java [hunk 6]
```diff
@@ -89,6 +96,71 @@ public Gem(final Storage storage) {
...
+                    | ExecutionException exc) {
+                    throw new ArtipieException(exc);
+                }
+                return filecontent;
+            }
+        ).handle(removeTempDir(dir.get()));
+    }
+
+    /**
+     * Get info Ruby gem.
```
Comment (g4s8) [inline src/main/java/com/artipie/gem/Gem.java:137]: Two same methods?

File: src/main/java/com/artipie/gem/Gem.java [hunk 7]
```diff
@@ -89,6 +96,71 @@ public Gem(final Storage storage) {
...
+     *
+     * @param filename Ruby file to return
+     * @return Completable action
+     */
+    public CompletionStage<byte[]> getRubyFile(final Key filename) {
+        final AtomicReference<Path> dir = new AtomicReference<>();
+        return newTempDir().thenCompose(
+            tmp -> {
+                dir.set(tmp);
+                return new Copy(this.storage, key -> key.string().contains(filename.string()))
```
Comment (g4s8) [inline src/main/java/com/artipie/gem/Gem.java:110]: `contains` is not correct here because gem name could be in the repository name, e.g. user `rails` may keep package `gviz` in his repository, then the key will be `rails/gviz`

File: src/main/java/com/artipie/gem/Gem.java [hunk 8]
```diff
@@ -89,6 +96,71 @@ public Gem(final Storage storage) {
...
+     *
+     * @param gems Ruby gem to extract info
+     * @return Completable action
+     */
+    public CompletionStage<byte[]> getDependencies(final List<Key> gems) {
+        final AtomicReference<Path> dir = new AtomicReference<>();
+        return newTempDir().thenCompose(
+            tmp -> {
+                dir.set(tmp);
+                return new Copy(this.storage, key -> key.string().contains(gems.get(0).string()))
```
Comment (g4s8) [inline src/main/java/com/artipie/gem/Gem.java:147]: Why we copy only first gem here? And same problem for `contains`

File: src/main/java/com/artipie/gem/Gem.java [hunk 9]
```diff
@@ -157,6 +229,42 @@ public Gem(final Storage storage) {
...
     }
 
+    /**
+     * Find gem in a given path.
+     * @param gem Gem name to get info
+     * @param exact Is exact
+     * @param fallout Is fallout
+     * @return String full path to gem file
+     */
+    private CompletionStage<Key> getGemFile(final Key gem, final boolean exact,
```
Comment (g4s8) [inline src/main/java/com/artipie/gem/Gem.java:239]: The javadoc and method name are confusing here: `getGemFile` returns `Key`, not file. "Find gem in a given path." accepts `Key` not path.

File: src/main/java/com/artipie/gem/Gem.java [hunk 10]
```diff
@@ -157,6 +229,42 @@ public Gem(final Storage storage) {
         );
     }
 
+    /**
+     * Find gem in a given path.
+     * @param gem Gem name to get info
+     * @param exact Is exact
+     * @param fallout Is fallout
```
Comment (g4s8) [inline src/main/java/com/artipie/gem/Gem.java:236]: What is fallout? Method behavior should be clear from Javadoc description

File: src/main/java/com/artipie/gem/Gem.java [hunk 11]
```diff
@@ -157,6 +229,42 @@ public Gem(final Storage storage) {
...
+    private CompletionStage<Key> getGemFile(final Key gem, final boolean exact,
+        final Key fallout) {
+        final String gemstr = ".gem";
+        final CompletableFuture<Key> future = new CompletableFuture<>();
+        Single.fromFuture(this.storage.list(Key.ROOT))
+            .map(
+                list -> {
+                    final List<Key> result;
+                    final Stream<Key> res = list.stream().filter(
+                        key -> key.string().contains(gem.string()) && key.string().endsWith(gemstr)
```
Comment (g4s8) [inline src/main/java/com/artipie/gem/Gem.java:248]: Same here: we can match `rails/gviz.gem` where gem is `rails`

File: src/main/java/com/artipie/gem/Gem.java [hunk 12]
```diff
@@ -157,6 +229,42 @@ public Gem(final Storage storage) {
...
+                            key -> key.string().contains(gem.string())
+                                && key.string().endsWith(gemstr)
+                        ).limit(1).collect(Collectors.toList());
+                    } else {
+                        result = list.stream().filter(
+                            key -> !exact && key.string().contains(gem.string())
+                                || fallout != null && key.string().equals(fallout.string())
+                        ).limit(1).collect(Collectors.toList());
+                    }
+                    return result;
```
Comment (g4s8) [inline src/main/java/com/artipie/gem/Gem.java:261]: This code has too many duplications and perform the same work multiple times - no need to build a stream just to count and then build another stream for filter. 
Also, it's not clear why do you use `List result` which is always containing only one element.

File: src/main/java/com/artipie/gem/http/ApiGetSlice.java [hunk 13]
```diff
@@ -75,18 +91,81 @@ public ApiGetSlice(final Storage storage) {
     public Response response(final String line,
         final Iterable<Map.Entry<String, String>> headers,
         final Publisher<ByteBuffer> body) {
+        final String deproute = "/api/v1/dependencies";
```
Comment (g4s8) [inline src/main/java/com/artipie/gem/http/ApiGetSlice.java:94]: No need to put all HTTP handlers into one `Slice`, you can introduce more slices to handle new endpoints. One slice per endpoint

File: src/main/java/com/artipie/gem/http/ApiGetSlice.java [hunk 14]
```diff
@@ -75,18 +91,81 @@ public ApiGetSlice(final Storage storage) {
...
+                    .thenApply(out -> new RsWithBody(ByteBuffer.wrap(out)))
+            );
+        } else {
             throw new IllegalStateException("Invalid routing schema");
         }
+        return res;
+    }
+
+    /**
+     * Process dependency route.
```
Comment (g4s8) [inline src/main/java/com/artipie/gem/http/ApiGetSlice.java:144]: Please use meaningful method names and javadocs. Try to imagine that you see this code for the first time, how would you understand what does this method do?

File: src/main/java/com/artipie/gem/http/ApiGetSlice.java [hunk 15]
```diff
@@ -75,18 +91,81 @@ public ApiGetSlice(final Storage storage) {
...
+        final int indexs = line.indexOf(deproute) + offset;
+        final int indexe = line.indexOf(ApiGetSlice.HTTP) - 1;
+        final List<Key> gemkeys = new ArrayList<>(0);
+        for (final String gemname : line.substring(indexs, indexe).split(",")) {
+            gemkeys.add(new Key.From(gemname));
+        }
+        final byte[] obj;
+        try {
+            obj = this.sdk.getDependencies(gemkeys)
+                .toCompletableFuture().get();
```
Comment (g4s8) [inline src/main/java/com/artipie/gem/http/ApiGetSlice.java:161]: Don't ever use blocking calls. It's only acceptable in unit tests or some corner cases like futures joining

File: src/main/java/com/artipie/gem/http/ApiGetSlice.java [hunk 16]
```diff
@@ -75,18 +91,81 @@ public ApiGetSlice(final Storage storage) {
...
+        } catch (final InterruptedException | ExecutionException exc) {
```
Comment (g4s8) [inline src/main/java/com/artipie/gem/http/ApiGetSlice.java:162]: `InterruptedException` should not be handled in a such way, please check documentation. You can find a very good summary from documentation here: https://stackoverflow.com/a/3976377/1723695

File: src/main/java/com/artipie/gem/http/GemSlice.java [hunk 17]
```diff
@@ -95,6 +95,48 @@ public GemSlice(final Storage storage,
...
+                    ),
+                    new ApiGetSlice(storage)
+                ),
+                new RtRulePath(
+                    new RtRule.All(
+                        new ByMethodsRule(RqMethod.GET),
+                        new RtRule.ByPath("/gems/.*")
+                    ),
+                    new ApiGetSlice(storage)
+                ),
```
Comment (g4s8) [inline src/main/java/com/artipie/gem/http/GemSlice.java:139]: Same problem here: no need to handle all endpoints in one slice

File: src/main/java/com/artipie/gem/ruby/IDependencies.java [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+ * SOFTWARE.
+ */
+package com.artipie.gem.ruby;
+
+/**
+ * Ruby Dependencies gem index implementation.
+ *
+ * @since 1.0
+ */
+public interface IDependencies {
```
Comment (g4s8) [inline src/main/java/com/artipie/gem/ruby/IDependencies.java:31]: 1) why do we need this interface? 2) what does this `I` prefix mean?

File: src/main/ruby/dependencies.rb [hunk 19]
```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+require 'rubygems'
+require 'rubygems/package'
+require 'time'
+require 'tmpdir'
+require 'java'
+
+class Dependencies
+    include com.artipie.gem.ruby.IDependencies
```
Comment (g4s8) [inline src/main/ruby/dependencies.rb:9]: No need to implement some interfaces here, we don't expose it outside the package. You can just call class method directly as we did previously

File: src/main/ruby/dependencies.rb [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+
+    def initialize(val)
+        @val = val
+    end
+
+    def dependencies()
+        resgems = []
+        @val.split.each do |gem|
+            puts(gem)
```
Comment (g4s8) [inline src/main/ruby/dependencies.rb:18]: Please remove stdout logging from production code, you can use only `Logger` from slf4j

**Review by g4s8 — CHANGES_REQUESTED** [review]
@milkysoft please check some initial comments

File: src/main/ruby/dependencies.rb (hunk 20 above)
Comment (milkysoft) [reply src/main/ruby/dependencies.rb]: @g4s8 , agreed, fixed in latest commit

File: src/main/java/com/artipie/gem/Gem.java (hunk 4 above)
Comment (milkysoft) [reply src/main/java/com/artipie/gem/Gem.java]: @g4s8 , agreed, fixed in latest commit

File: src/main/java/com/artipie/gem/Gem.java (hunk 3 above)
Comment (milkysoft) [reply src/main/java/com/artipie/gem/Gem.java]: @g4s8 , agreed, fixed in latest commit

File: src/main/java/com/artipie/gem/http/GemSlice.java (hunk 17 above)
Comment (milkysoft) [reply src/main/java/com/artipie/gem/http/GemSlice.java]: @g4s8 , agreed, fixed in latest commit

File: src/main/java/com/artipie/gem/Gem.java (hunk 1 above)
Comment (milkysoft) [reply src/main/java/com/artipie/gem/Gem.java:97]: @g4s8 , we use the same functionality for all responses, althoough we can reconsider our approach

File: src/main/java/com/artipie/gem/Gem.java (hunk 2 above)
Comment (milkysoft) [reply src/main/java/com/artipie/gem/Gem.java:102]: @g4s8 , I confirm **Key** here is arbitrary name

File: src/main/ruby/dependencies.rb (hunk 19 above)
Comment (milkysoft) [reply src/main/ruby/dependencies.rb:6]: @g4s8 , unfortunately all efforts to convert RubyObject to RubyString failed until now

Comment (milkysoft) [reply src/main/ruby/dependencies.rb]: @g4s8 , fixed in latest commit

File: src/main/java/com/artipie/gem/ruby/IDependencies.java (hunk 18 above)
Comment (milkysoft) [reply src/main/java/com/artipie/gem/ruby/IDependencies.java]: @g4s8 , fixed in latest commit

File: src/main/java/com/artipie/gem/Gem.java (hunk 2 above)
Comment (g4s8) [reply src/main/java/com/artipie/gem/Gem.java:102]: @milkysoft but `key` is just a locator of some data in key-value storage, it could be a file or could not
